File: Wizard Battle Web Client/src/app/services/transaction.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { catchError, Observable, ObservableInput } from 'rxjs';
import { observableToBeFn } from 'rxjs/internal/testing/TestScheduler';
import { environment } from 'src/environments/environment';
import { DirectTransactionResponse, StaticTransactionResponse, TransactionRequest } from '../_models/Transaction';

@Injectable({
  providedIn: 'root'
})
export class TransactionService {

	private url: string = environment.ApiUrl + "/Transaction";

  	constructor(private http: HttpClient) { }

	public getAll(): Observable<StaticTransactionResponse[]>
	{
		return this.http.get<StaticTransactionResponse[]>(this.url).pipe(catchError<StaticTransactionResponse[],ObservableInput<StaticTransactionResponse[]>>(Response=>{
			var a: StaticTransactionResponse[]=[]
			 	  if(Response == null)
			 	  {
			 		  return a
					
			 	  }
			 	return Response
			   }))
	}

	public getById(id: number): Observable<DirectTransactionResponse>
	{
	  return this.http.get<DirectTransactionResponse>(`${this.url}/${id}`);
	}

	public create(request: TransactionRequest): Observable<DirectTransactionResponse>
	{
	  return this.http.post<DirectTransactionResponse>(this.url, request);
	}

	public update(id: number, request: TransactionRequest): Observable<DirectTransactionResponse>
	{
	  return this.http.put<DirectTransactionResponse>(`${this.url}/${id}`, request);
	}

	public delete(id: number): Observable<DirectTransactionResponse>
	{
	  return this.http.delete<DirectTransactionResponse>(`${this.url}/${id}`);
	}
}

File: Wizard Battle Web API/Services/TransactionService.cs
namespace Wizard_Battle_Web_API.Services
{
	public interface ITransactionService
	{
		public Task<List<StaticTransactionResponse>> GetAll();
		public Task<DirectTransactionResponse> GetById(int transactionId);
		public Task<DirectTransactionResponse> Create(TransactionRequest request);
		public Task<DirectTransactionResponse> Update(int transactionId, TransactionRequest request);
		public Task<DirectTransactionResponse> Delete(int transactionId);
	}
	public class TransactionService : ITransactionService
	{

		private readonly ITransactionRepository m_transactionRepository;
		private readonly IMapper m_mapper;


		public TransactionService(ITransactionRepository transactionRepository, IMapper mapper)
		{
			m_transactionRepository = transactionRepository;
			m_mapper = mapper;
		}


		public async Task<DirectTransactionResponse> Create(TransactionRequest request)
		{
			Transaction transaction = await m_transactionRepository.Create(m_mapper.Map<Transaction>(request));
			if(transaction != null)
			{
				return m_mapper.Map<DirectTransactionResponse>(transaction);
			}

			return null;
		}


		public async Task<DirectTransactionResponse> Delete(int transactionId)
		{
			Transaction transaction = await m_transactionRepository.GetById(transactionId);
			if(transaction != null)
			{
				return m_mapper.Map<DirectTransactionResponse>(await m_transactionRepository.Delete(transaction));
			}

			return null;
		}


		public async Task<List<StaticTransactionResponse>> GetAll()
		{
			List<Transaction> transactions = await m_transactionRepository.GetAll();
			if(transactions != null)
			{
				return transactions.Select(x => m_mapper.Map<StaticTransactionResponse>(x)).ToList();
			}

			return null;
		}


		public async Task<DirectTransactionResponse> GetById(int transactionId)
		{
			Transaction transaction = await m_transactionRepository.GetById(transactionId);
			if(transaction != null)
			{
				return m_mapper.Map<DirectTransactionResponse>(transaction);
			}

			return null;
		}


		public async Task<DirectTransactionResponse> Update(int transactionId, TransactionRequest request)
		{
			Transaction transaction = await m_transactionRepository.Update(transactionId, m_mapper.Map<Transaction>(request));
			if(transaction != null)
			{
				return m_mapper.Map<DirectTransactionResponse>(transaction);
			}

			return null;
		}
	}
}
